### ISSUE TITLE
Solving wonky data problem after doing nsampstp = 2*tau

### ISSUE THREAD

**Issue by kunal1729verma** [kunal1729verma/ising-2d#1]
So if I run the monte carlo function with `nsampstp` = $2 \tau$, the graphs don't come out as nice as I had expected, probably because (this is my guess) the number of measurements = floor(`nsamsweeps`/`nsampstp`) where `nsamsweeps` remains same (say 50000). Since `nsampstp` can vary anywhere from 10 to 500, the number of measurements aren't consistent, and therefore the averages will be computed over datasets with varying length.

To fix this, I thought of putting `nsamsweeps` = $2\cdot \tau \cdot 5000$ to keep the number of measurements consistent (5000), but that's again producing those nasty `nan` in columns 7, 8, 9, 10, 13, 14 (starting from 0 as in my notes). I've somehow gotta figure out the source of these `nan` errors if I want to move on further.

**Comment by kunal1729verma** [comment]
Some success. After hours of troubleshooting, I have determined that the `nan` error occurs whenever `observables.size()`%`nblen`  = 1. Now why that happens, I'm trying to investigate. Will get back to it soon.

**Comment by kunal1729verma** [comment]
Just a general comment on how I recognized the issue:

(1) The measurements were always perfectly fine. All values were real, no infinities or blowing up or divide by zero. The `nan` error always appeared after processing it through the binned mean and binned variance function.

(2) The error didn't occur always. In a set of 25 temperature iterations, it rarely occurred once, not even once most of the times. This pointed towards the error coming up with a probability close to 1/25. What could cause it? My guess was whenever `observables.size()`%`nblen` = 0, some weird mishap happened because then the error would come up with a probability of 1/20. So I tried to investigate if that was the case.

Fortunately, my guess worked, but turns out the error occurs when `observables.size()`%`nblen` = 1, not 0.

Now I shall investigate why that's happening.

**Comment by kunal1729verma** [comment]
FOUND THE NAN ERROR! 

Whenever `observables.size()`%`nblen` = 1, then on the step when `idat == ndat`, => ibin = 1. This causes the variable `factor` = `nan` because we divide by zero!

**Comment by kunal1729verma** [comment]
Replaced

```
double factor;
factor = sqrt((double(ibin))/((double)(ibin-1))) ;
```

with

```
double factor;
if (ibin == 1) {
  factor = 0;
}
else {
    factor = sqrt((double(ibin))/((double)(ibin-1))) ;   
}
```

**Comment by kunal1729verma** [comment]
![image](https://user-images.githubusercontent.com/69515542/170583917-1ff7f007-3c9a-49ea-a39c-d3f592b79a71.png)
![image](https://user-images.githubusercontent.com/69515542/170583958-b6f68776-422b-44b0-b49b-3871b053dab5.png)

Woohooo!! It's working!

**Comment by kunal1729verma** [comment]
Now, we move to-

(1) checking if the correlation times are even sensible.
(2) find a better way to get better data without just ramping up the system size L.
(3) should `nsamsweeps` really depend on $\tau$?
(4) write the code snippet to find critical exponents directly in ising.cpp, or maybe use the _out.plt data and do the critical exponent analysis in python.

**Comment by kunal1729verma** [comment]
Okay, so it seems like the way I was calculating autocorrelation time wasn't the correct way. It's not a good idea to sum up all the values till infinity because after some time, the correlation function tends to be dominated by statistical errors more than correlations themselves. Therefore, we use the "automatic-windowing" algorithm to deal with this issue. 

First of all, start with a large enough $t_\text{max}$. The upper limit of your sum shouldn't exceed the value of $t_\text{max}$, otherwise you will miss adding up the actual correlations. So choose a high value sich as $t_\text{max} = 2000$ and then see what's the maximum value of the upper limit that can reach and keep $t_\text{max}$ at a safe value above that limit.

Now, choose $M$ to be the smallest integer such that $M \geq 6\tau_\text{int} (M)$. To elaborate the procedure, the algorithm is as follows - 
1. Start with some small value of $M$, let's say $M = 50$.

2. Evaluate $\tau_\text{int} (M)$ by using $M = 50$ as the cutoff in the sum
$$\tau_\text{int} (M) = \frac{1}{2} + \sum_{i=1}^{M} A_Q(i)$$

3. Check if $M \geq 6\tau_\text{int} (M)$, or in this specific example, if $50 \geq 6 \tau_\text{int}(50)$:
 a. If  $M < 6\tau_\text{int} (M)$, increase the value of $M$, let's say double it, and repeat step 2 and 3.
 b. If $M \geq 6\tau_\text{int} (M)$ is satisfied, then we use that value of $M = M^\star$ as the upper limit and the correct correlation estimate is given by $\tau_\text{int} (M^\star)$.

**Comment by kunal1729verma** [comment]
As of now, the code is _supposed_ to work properly, but apparently, it's not. As far as I understand, the reason why I'm not getting good quality data (and the reason why it's taking so much time) is that the calculated correlation times are incorrect. The method of calculating the correlation time is correct, but the underlying correlation function values $A_Q(i)$'s are not correct, and I'm sure about that. One of the graphs for T=2.02 illustrates that. The correlation function, by definition, shouldn't go outside the range $[-1,1]$, and here it drops below $-200$. 

![t=2 02](https://user-images.githubusercontent.com/69515542/173029969-5d87ec2f-46f9-4ac1-b518-78966b689e35.png)

Therefore, the almost final step to getting everything working correctly is to see what's causing the correlation function to output such hideous numbers. I need to fix it asap.

**Comment by kunal1729verma** [comment]
The wonky data problem is more or less solved. I'll just drop my discord update here

finally, i can generate some reliable data from the cpp program. the reason why I was getting wonky data i guess was because I wasn't averaging over enough measurements. the above physical quantities are derived on taking averages over 20000 measurements. 

now, there are two more small tasks left - 
(1) find a reliable method to calculate errors. the errors being calculated in the current program are way too high.
(2) find a way to implement autocorrelation calculations via FFT.
(3) push the cpp file to github.
(4) finally, find the appropriate argument for using |m| in place of m.

once I have a reliable way of calculating the errors as well, and hopefully if I find a way to implement autocorr calcs using FFT, then the code can quickly give me data for the derived physical quantities which I can use for finite size scaling methods.

**Comment by kunal1729verma** [comment]
some graphs for L = 16 from present data
![image](https://user-images.githubusercontent.com/69515542/174901576-08b4c858-1abe-4159-93a3-473788870b87.png)
![image](https://user-images.githubusercontent.com/69515542/174901605-94a4d217-8fdc-43f4-9a22-be25a27be427.png)
![image](https://user-images.githubusercontent.com/69515542/174901689-a9aba59e-f6f3-4930-b279-67cdd277e631.png)

**Comment by kunal1729verma** [comment]
So I was finally able to solve this issue by computing Jackknife errors for all the observable expectation values and derived quantities, and the graphs look much more sensible and coherent. I was also able to calculate the critical exponents using the data of $L = 8, 16, 32$, and I got pretty decent results. The only problems I faced were - 
1. Calculation of errors in the estimated critical exponents?
2. Incredibly long runtimes.

For 2, the conclusion is as follows - autocorrelation analysis doesn't take much time; instead, the main Monte Carlo loop is the most expensive part of the program. 99% of the runtime is spent doing Monte Carlo spin flips. The solution to this issue is - 
a. Parallelize this code.
b. Use Cluster Algorithms.

When combined, I'm pretty sure we're gonna get impressive speed-ups. And that'll be it for the Classical spin Ising Model.
